### ISSUE TITLE
Conditional padding issue with jit_compile=True

### ISSUE THREAD

**Issue by shkarupa-alex** [tensorflow/tensorflow#59474]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Have you reproduced the bug with TF nightly?

Yes

### Source

source

### Tensorflow Version

2.9.2

### Custom Code

Yes

### OS Platform and Distribution

Google Colab

### Python version

Google Colab

### Current Behaviour?

```shell
Function smart_cond (tf.cond) fails to pass gradients back when combined with padding/slicing if JIT enabled.
```

### Standalone code to reproduce the issue

```shell
https://colab.research.google.com/drive/18O-DM1vEYbmOhQ50zt6f5KVy-8qbMqBH?usp=sharing
```

### Relevant log output

```shell
InvalidArgumentError: Graph execution error:

Incompatible shapes: [16,8,12,3] vs. [16,8,10,3]

Stack trace for op definition: 
File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
  return _run_code(code, main_globals, None,
File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
  exec(code, run_globals)
File "/usr/local/lib/python3.8/dist-packages/ipykernel_launcher.py", line 16, in <module>
  app.launch_new_instance()
File "/usr/local/lib/python3.8/dist-packages/traitlets/config/application.py", line 992, in launch_instance
  app.start()
File "/usr/local/lib/python3.8/dist-packages/ipykernel/kernelapp.py", line 612, in start
  self.io_loop.start()
File "/usr/local/lib/python3.8/dist-packages/tornado/platform/asyncio.py", line 149, in start
  self.asyncio_loop.run_forever()
File "/usr/lib/python3.8/asyncio/base_events.py", line 570, in run_forever
  self._run_once()
File "/usr/lib/python3.8/asyncio/base_events.py", line 1859, in _run_once
  handle._run()
File "/usr/lib/python3.8/asyncio/events.py", line 81, in _run
  self._context.run(self._callback, *self._args)
File "/usr/local/lib/python3.8/dist-packages/tornado/ioloop.py", line 690, in <lambda>
  lambda f: self._run_callback(functools.partial(callback, future))
File "/usr/local/lib/python3.8/dist-packages/tornado/ioloop.py", line 743, in _run_callback
  ret = callback()
File "/usr/local/lib/python3.8/dist-packages/tornado/gen.py", line 787, in inner
  self.run()
File "/usr/local/lib/python3.8/dist-packages/tornado/gen.py", line 748, in run
  yielded = self.gen.send(value)
File "/usr/local/lib/python3.8/dist-packages/ipykernel/kernelbase.py", line 381, in dispatch_queue
  yield self.process_one()
File "/usr/local/lib/python3.8/dist-packages/tornado/gen.py", line 225, in wrapper
  runner = Runner(result, future, yielded)
File "/usr/local/lib/python3.8/dist-packages/tornado/gen.py", line 714, in __init__
  self.run()
File "/usr/local/lib/python3.8/dist-packages/tornado/gen.py", line 748, in run
  yielded = self.gen.send(value)
File "/usr/local/lib/python3.8/dist-packages/ipykernel/kernelbase.py", line 365, in process_one
  yield gen.maybe_future(dispatch(*args))
File "/usr/local/lib/python3.8/dist-packages/tornado/gen.py", line 209, in wrapper
  yielded = next(result)
File "/usr/local/lib/python3.8/dist-packages/ipykernel/kernelbase.py", line 268, in dispatch_shell
  yield gen.maybe_future(handler(stream, idents, msg))
File "/usr/local/lib/python3.8/dist-packages/tornado/gen.py", line 209, in wrapper
  yielded = next(result)
File "/usr/local/lib/python3.8/dist-packages/ipykernel/kernelbase.py", line 543, in execute_request
  self.do_execute(
File "/usr/local/lib/python3.8/dist-packages/tornado/gen.py", line 209, in wrapper
  yielded = next(result)
File "/usr/local/lib/python3.8/dist-packages/ipykernel/ipkernel.py", line 306, in do_execute
  res = shell.run_cell(code, store_history=store_history, silent=silent)
File "/usr/local/lib/python3.8/dist-packages/ipykernel/zmqshell.py", line 536, in run_cell
  return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)
File "/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py", line 2854, in run_cell
  result = self._run_cell(
File "/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py", line 2881, in _run_cell
  return runner(coro)
File "/usr/local/lib/python3.8/dist-packages/IPython/core/async_helpers.py", line 68, in _pseudo_sync_runner
  coro.send(None)
File "/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py", line 3057, in run_cell_async
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
File "/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py", line 3249, in run_ast_nodes
  if (await self.run_code(code, result,  async_=asy)):
File "/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py", line 3326, in run_code
  exec(code_obj, self.user_global_ns, self.user_ns)
File "<ipython-input-5-0e949eda0b3b>", line 2, in <module>
  model.fit(
File "/usr/local/lib/python3.8/dist-packages/keras/utils/traceback_utils.py", line 64, in error_handler
  return fn(*args, **kwargs)
File "/usr/local/lib/python3.8/dist-packages/keras/engine/training.py", line 1409, in fit
  tmp_logs = self.train_function(iterator)
File "/usr/local/lib/python3.8/dist-packages/keras/engine/training.py", line 1051, in train_function
  return step_function(self, iterator)
File "/usr/local/lib/python3.8/dist-packages/keras/engine/training.py", line 1040, in step_function
  outputs = model.distribute_strategy.run(run_step, args=(data,))
File "/usr/local/lib/python3.8/dist-packages/keras/engine/training.py", line 1030, in run_step
  outputs = model.train_step(data)
File "/usr/local/lib/python3.8/dist-packages/keras/engine/training.py", line 893, in train_step
  self.optimizer.minimize(loss, self.trainable_variables, tape=tape)
File "/usr/local/lib/python3.8/dist-packages/keras/optimizers/optimizer_v2/optimizer_v2.py", line 537, in minimize
  grads_and_vars = self._compute_gradients(
File "/usr/local/lib/python3.8/dist-packages/keras/optimizers/optimizer_v2/optimizer_v2.py", line 590, in _compute_gradients
  grads_and_vars = self._get_gradients(tape, loss, var_list, grad_loss)
File "/usr/local/lib/python3.8/dist-packages/keras/optimizers/optimizer_v2/optimizer_v2.py", line 471, in _get_gradients
  grads = tape.gradient(loss, var_list, grad_loss)

	 [[{{node gradient_tape/mean_squared_error/BroadcastGradientArgs}}]]
	 [[StatefulPartitionedCall]] [Op:__inference_train_function_1464]
```
</details>

**Comment by SuryanarayanaY** [comment]
I tried the code provided by you and it works fine without jit_compile. However with `jit_compile=True` the code fails with the graph execution error as reported.Please refer to attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/256b4aae6bc75b30500646502b77886d/59474_2-9-2v.ipynb). 
 
@shkarupa-alex ,

The reason for the error might be due to the fact that present all TF Ops and Operations are not compilable with `jit_compile`. lambda variables not working as intended within tensorflow loops.For more details on this please refer this [source](https://github.com/tensorflow/tensorflow/issues/56089).This might be the reason for the error.

Hence I tried `autoclustering` by setting `tf.config.optimizer.set_jit("autoclustering" )`  and code executed fine. Please refer to attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/09a11142ccf9b8d91e95bdd8a7cb0577/59474_2-11-auto_clustering.ipynb) with `autoclustering`. Auto clustering automatically enables `jit_compile=True` for those which are supported with `jit_compile` and ignores those if not supported.I Hope this may be of some help.As there is no error with auto_clustering means there are some operations not supported with `jit_compile=True`. The best way is use `autoclustering` or custom training loop with those operations that supports jit compilation kept under `@tf.function(jit_compile=True)`. Thanks!

**Comment by shkarupa-alex** [comment]
According to this doc https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2xla/g3doc/gpu_supported_ops.md both Pad and Slice ops support XLA

**Comment by shkarupa-alex** [comment]
Also i figured out that issue comes from slicing (second smart_cond in my example)

**Comment by kiransair** [comment]
Hi @shkarupa-alex, I did not face any error while using `jit_compile=True` in colab. please refer to this [gist](https://colab.research.google.com/gist/kiransair/31d58a1013b9649f7ac26953ef325f10/tf_forum_59474.ipynb). Thank You.

**Comment by shkarupa-alex** [comment]
Seems to be fixed since v2.14
